src/libraries/System.Net.Security/src/System/Net/Security/SecureChannel.cs

```diff
@@ -948,38 +964,48 @@ internal bool VerifyRemoteCertificate(RemoteCertValidationCallback? remoteCertVa
                     sslPolicyErrors |= CertificateValidationPal.VerifyCertificateProperties(
                         _securityContext!,
                         chain,
-                        remoteCertificateEx,
+                        _remoteCertificate,
                         _sslAuthenticationOptions.CheckCertName,
                         _sslAuthenticationOptions.IsServer,
                         _sslAuthenticationOptions.TargetHost);
                 }
 
                 if (remoteCertValidationCallback != null)
                 {
-                    success = remoteCertValidationCallback(_sslAuthenticationOptions.TargetHost, remoteCertificateEx, chain, sslPolicyErrors);
+                    object? sender = _ssl;
```

Comment:
What's this check needed for?

Also, it seems like the only place you use `_ssl` is for this check. If you're just using it to determine whether the SecureChannel is disposed, could you just use `bool _disposed` instead?

Reply:
I was worried about race conditions. Since there is no lock, grabbing global value and testing it seemed like safest.
If I use _disposed, it can be modified between check and the passing it to the callback, right?

Reply:
Ah, I see, you're passing it to the validation callback below. Ok, makes sense.

Reply:
Here's my suggestion: Don't store the SslStream in `_ssl`. Just pass it into `VerifyRemoteCertificate` -- that seems to be the only place it's used.

Reply:
I was thinking about it but I was not sure if it is OK to pass `null` to that callback. I think we would not do that in the past. 
It would certainly simplify this. 

In the past SecureChannel did not have any real clue about SslStream - and that make tracing more difficult as we deal with objects caller has no visibility to. This wrapper business is yet another case. (I did not eliminated all the cases - just the one I cared about for the remote certificate as that is most common people hit)
I was thinking about merging them to one - but that is bigger change that we should do right now. (and it may have other issues)

Reply:
If you pass SslStream to this routine, then why would it be null? I'm not sure exactly where this is called from, but I guess I assumed it was from SslStream?

> I was thinking about merging them to one - but that is bigger change that we should do right now. (and it may have other issues)

Yep, merging seems good, but not right now.

Reply:
it is called from within SslStream. But either with sync or async, somebody can get tired of waiting and call Dispose from other thread and move one. Maybe I'm just paranoid here and if somebody does, possible NRE inside of the callback would be acceptable. Our code really does not care AFAIK. 
I was also thinking about just leaving `_ssl` on Dispose but I wanted to cut the circular reference.

Reply:
If you pass the SslStream into this function, then they shouldn't get NRE, though they could get ODE. But I think that's fine, and unavoidable anyway since it's a race.